### PR TITLE
Adding Defiler upgrades

### PR DESCRIPTION
**Pull request by EnvyDragon**
Needs generator change too, will do that PR next.

Upgrades:
- Pathogen Projectors: +4 range on Plague and Dark Swarm.
- Trapdoor Adaptation: Defilers can cast while burrowed.
- Predatory Consumption: Defilers can cast Consume on any non-heroic biological target, not just friendly Zerg.
- Comorbidity: Plague can stack up to 3 times and depletes energy as well as health.

### CODE REVIEW

**Comment by EnvyDragon** [comment]
Generator PR here: https://github.com/Ziktofel/Archipelago/pull/200

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/AbilData.xml [hunk 1]
```diff
@@ -10149,6 +10147,21 @@
...
+        <TargetFilters value="Biological;Self,Enemy,Structure,Heroic,Missile,Uncommandable,Stasis,Dead,Hidden,Hallucination,Invulnerable"/>
+        <Range value="4"/>
+        <Effect index="0" value="AP_DefilerConsumption"/>
+        <CmdButtonArray index="Execute" DefaultButtonFace="AP_DefilerConsumption" Requirements="AP_NotHaveDefilerPredatoryConsumption"/>
+    </CAbilEffectTarget>
+    <CAbilEffectTarget id="AP_DefilerSCBWPredatoryConsumption" parent="AP_DefilerSCBWConsumption">
+        <TargetFilters index="0" value="Biological;Self,Structure,Heroic,Missile,Uncommandable,Stasis,Dead,Hidden,Hallucination,Invulnerable"/>
+        <Effect index="0" value="AP_DefilerPredatoryConsumption"/>
+        <CmdButtonArray index="Execute" DefaultButtonFace="AP_DefilerConsumption" Requirements="AP_HaveDefilerPredatoryConsumption"/>
+    </CAbilEffectTarget>
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/AbilData.xml:10164]: lacks range actor:
Snímek obrazovky pořízený 2024-05-19 11-56-06
compared to this:
Snímek obrazovky pořízený 2024-05-19 11-59-13
Also the graphic effect/sound is missing (See actors - check references towards Infestor consume)

File: Mods/ArchipelagoPlayer.SC2Mod/enUS.SC2Data/LocalizedData/GameStrings.txt [hunk 2]
```diff
@@ -1866,9 +1872,14 @@ Button/Tooltip/AP_DarkTemplarTaldarim=Tal'darim Faction<n/>Deadly warrior-assass
...
 Button/Tooltip/AP_DefilerSCBW=Subversion specialist. Able to manipulate battlefield conditions. Can use the Dark Swarm, Consume and Plague abilities.
 Button/Tooltip/AP_DefilerSCBWDarkSwarm=Creates a micro-organism cloud preventing all incoming ranged damage for <d ref="Effect,AP_DefilerSCBWDarkSwarm@CP,PeriodCount*Effect,AP_DefilerSCBWDarkSwarm@CP,PeriodicPeriodArray[0]"/> seconds to ground units in the area. Doesn't protect against splash damage type or spells.
 Button/Tooltip/AP_DefilerSCBWPlague=Infects all units and buildings in the target area, inflicting <d ref="Effect,AP_DefilerSCBWPlague@Damage,Amount*Behavior,AP_DefilerSCBWPlague,Duration/Behavior,AP_DefilerSCBWPlague,Period"/> damage over <d ref="Behavior,AP_DefilerSCBWPlague,Duration"/> seconds. Cannot reduce units or buildings below 1 life. Ignores shields. Reveals infected cloaked and burrowed units.
+Button/Tooltip/AP_DefilerSCBWImprovedCastRange=Increases the cast range of Dark Swarm and Plague by <d ref="Upgrade,AP_ViperImprovedCastRange,EffectArray[0].Value"/>.
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/enUS.SC2Data/LocalizedData/GameStrings.txt:1879]: AP_DefilerImprovedCastRange

**Review by Ziktofel — CHANGES_REQUESTED** [review]

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/AbilData.xml (hunk 1 above)
Comment (EnvyDragon) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/AbilData.xml:10164]: Thanks for spotting that! Fixed.

File: Mods/ArchipelagoPlayer.SC2Mod/enUS.SC2Data/LocalizedData/GameStrings.txt (hunk 2 above)
Comment (EnvyDragon) [reply Mods/ArchipelagoPlayer.SC2Mod/enUS.SC2Data/LocalizedData/GameStrings.txt]: Good catch, fixed

**Comment by Ziktofel** [comment]
Snímek obrazovky pořízený 2024-05-20 00-19-44
Snímek obrazovky pořízený 2024-05-20 00-20-29
Plagued units donẗ have the visual effect

**Comment by EnvyDragon** [comment]
Went through every effect/behavior/ability that was added or modified and made sure there were actors for every case; it looked to me like Plague was the only one left that I missed, so I think it's good to go now?

**Comment by Ziktofel** [comment]
After I'll get to re-check everything, better be safe then sorry

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/ActorData.xml [hunk 3]
```diff
@@ -11551,17 +11553,23 @@
...
         <On Terms="Effect.AP_InfestorConsumption.Start; At Caster" Send="Create"/>
+        <On Terms="Effect.AP_DefilerConsumption.Start; At Caster" Send="Create"/>
+        <On Terms="Effect.AP_DefilerPredatoryConsumption.Start; At Caster" Send="Create"/>
         <Host Subject="_Selectable"/>
         <HostSiteOps Ops="SOpAttachCenter"/>
         <Model value="ConsumptionImpact"/>
     </CActorModel>
     <CActorRange id="AP_InfestorConsumptionRange" parent="RangeAbil" abil="AP_InfestorConsumption"/>
+    <CActorRange id="AP_DefilerConsumptionRange" parent="RangeAbil" abil="AP_DefilerConsumption"/>
+    <CActorRange id="AP_DefilerPredatoryConsumptionRange" parent="RangeAbil" abil="AP_DefilerPredatoryConsumption"/>
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/ActorData.xml:11572]: Typo: AP_DefilerSCBWConsumption (the Abil id, not the Effect's)

**Comment by Ziktofel** [comment]
BTW 
Snímek obrazovky pořízený 2024-05-20 19-41-28
Check the proper upgrade's effect index

**Comment by EnvyDragon** [comment]
test-screenshots
Tested and confirmed all strings are working; all actors related to modified abilities are confirmed working too.

And I tested for gameplay a while ago so I can confirm all the effects are working as intended, though it's hard to screenshot 1/2/3 damage+energy per tick DPS.

**Pull request merged**